[GStreamer] Implement pitch, rate, and volume controls of SpeechSynthesis

shivamidow opened the pull request:
#### a29a02ef293ea6f1711b746507f913f7a77a6e8d
<pre>
[GStreamer] Implement pitch, rate, and volume controls of SpeechSynthesis
<a href="https://bugs.webkit.org/show_bug.cgi?id=251486">https://bugs.webkit.org/show_bug.cgi?id=251486</a>

Reviewed by Philippe Normand.

This change adds pitch, rate, and volume controls support to web speech synthesis.
We use a GStreamer plugin of soundtouch for the pitch change of voice and control
the voice rate with gst_element_seek.

* Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp:
(WebCore::GstSpeechSynthesisWrapper::GstSpeechSynthesisWrapper):
(WebCore::GstSpeechSynthesisWrapper::speakUtterance):

Canonical link: <a href="https://commits.webkit.org/259949@main">https://commits.webkit.org/259949@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1b679a0383701d636714d29f4f93304b2037b2a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15487 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39279 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115651 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175750 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110372 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16976 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6712 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98661 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115314 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112230 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12894 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95874 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40453 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94769 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27518 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8719 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28870 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9253 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5438 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14873 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48416 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6877 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10800 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp:
```diff
@@ -95,8 +98,18 @@ GstSpeechSynthesisWrapper::GstSpeechSynthesisWrapper(Ref<PlatformSpeechSynthesiz
 
-    gst_bin_add_many(GST_BIN_CAST(m_pipeline.get()), m_src.get(), audioSink.get(), nullptr);
-    gst_element_link_pads_full(m_src.get(), "src", audioSink.get(), "sink", GST_PAD_LINK_CHECK_NOTHING);
+    m_pitchElement = makeGStreamerElement("pitch", nullptr);
```
Here we need to consider the case where `pitch` is not shipped by the runtime. You'd need to emit `WTFLogAlways()` explaining the pitch feature is disabled.

shivamidow replied:
Done.

philn commented on Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp:
```diff
@@ -136,6 +149,16 @@ void GstSpeechSynthesisWrapper::speakUtterance(RefPtr<PlatformSpeechSynthesisUtt
     m_utterance = WTFMove(utterance);
     webKitFliteSrcSetUtterance(WEBKIT_FLITE_SRC(m_src.get()), m_utterance->voice(), m_utterance->text());
+
+    // The pitch element does not handle the rate bigger than 1.0. We control
```
`m_pitchElement` might be null here, so needs a check :)

shivamidow replied:
Done.

philn reviewed (changes requested):
Thanks! I'll deploy the SDK tomorrow and let you know here so this PR can be updated and the CI should be green then.

shivamidow committed:
[GStreamer] Implement pitch, rate, and volume controls of SpeechSynthesis
https://bugs.webkit.org/show_bug.cgi?id=251486

Reviewed by Philippe Normand.

This change adds pitch, rate, and volume controls support to web speech synthesis.
We use a GStreamer plugin of soundtouch for the pitch change of voice and control
the voice rate with gst_element_seek.

* Source/WebCore/platform/gstreamer/PlatformSpeechSynthesizerGStreamer.cpp:
(WebCore::GstSpeechSynthesisWrapper::GstSpeechSynthesisWrapper):
(WebCore::GstSpeechSynthesisWrapper::speakUtterance):

Canonical link: https://commits.webkit.org/259949@main
